### PR TITLE
Add FixPermissionsCommand. Closes #397

### LINKED ISSUE #397: Move sqlite database to storage folder

### PR DESCRIPTION
**Pull request by jbrooksuk**
I can't remember why were we chmod`./public`, so it's removed. This will fix the `storage` directory too and only run on the `database` directory if needs be.

### CODE REVIEW

**Comment by joecohens** [comment]
:+1:

File: src/Console/Commands/FixPermissionsCommand.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+        $database = Config::get('database.default');
+
+        if ($database === 'sqlite') {
+            chmod(app_path('database') . '/', 755);
+            chmod(Config::get('database.sqlite.database'), 755);
+        }
+    }
+
+    protected function recursiveChmod($path, $mode = '0755') {
```
Comment (joecohens) [inline src/Console/Commands/FixPermissionsCommand.php]: docblock missing :)

Comment (jbrooksuk) [inline src/Console/Commands/FixPermissionsCommand.php]: Damn you ;)

**Comment by jbrooksuk** [comment]
Will squash.

**Comment by jbrooksuk** [comment]
Darn it.

**Comment by jbrooksuk** [comment]
Ok, now we can merge.

**Comment by GrahamCampbell** [comment]
Man, styleci is so sexy now! Thanks so much @joecohens!

File: src/Console/Commands/FixPermissionsCommand.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+     * @var string
+     */
+    protected $name = 'cachet:chmod';
+
+    /**
+     * The console command description.
+     *
+     * @var string
+     */
+    protected $description = 'Fixes file permissions.';
```
Comment (GrahamCampbell) [inline src/Console/Commands/FixPermissionsCommand.php]: probably not the best description

File: src/Console/Commands/FixPermissionsCommand.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    public function fire()
+    {
+        $this->recursiveChmod(app_path('storage').'/');
+
+        $database = Config::get('database.default');
+
+        if ($database === 'sqlite') {
+            chmod(app_path('database').'/', 755);
+            chmod(Config::get('database.sqlite.database'), 755);
```
Comment (GrahamCampbell) [inline src/Console/Commands/FixPermissionsCommand.php]: Don't read config here. Inject the actual result of config::get into the constructor.

File: src/Providers/ConsoleServiceProvider.php [hunk 4]
```diff
@@ -24,6 +26,10 @@ public function boot()
      */
     public function register()
     {
+        $this->app->singleton('CachetHQ\Cachet\Console\Commands\FixPermissionsCommand', function ($app) {
+            return new FixPermissionsCommand();
```
Comment (GrahamCampbell) [inline src/Providers/ConsoleServiceProvider.php]: here is where you can inject it

File: src/Console/Commands/FixPermissionsCommand.php (hunk 2 above)
Comment (jbrooksuk) [inline src/Console/Commands/FixPermissionsCommand.php]: "Fixes file and directory permissions."?

File: src/Console/Commands/FixPermissionsCommand.php (hunk 3 above)
Comment (jbrooksuk) [inline src/Console/Commands/FixPermissionsCommand.php]: Ok.

**Comment by jbrooksuk** [comment]
All done.

**Comment by jbrooksuk** [comment]
StyleCI = :heart_eyes:

**Comment by jbrooksuk** [comment]
Rebased.

**Comment by GrahamCampbell** [comment]
Just need to remove the unused facade import now :)

File: src/Console/Commands/FixPermissionsCommand.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+     * @param string $databasePath
+     * @param string $databaseDefault
+     *
+     * @return void
+     */
+    public function __construct($storageDirectory, $databaseDirectory, $databasePath, $databaseDefault)
+    {
+        $this->storageDirectory  = $storageDirectory;
+        $this->databaseDirectory = $databaseDirectory;
+        $this->databasePath      = $databasePath;
```
Comment (GrahamCampbell) [inline src/Console/Commands/FixPermissionsCommand.php]: Can we not align these please?

**Comment by jbrooksuk** [comment]
We're almost there!

File: src/Console/Commands/FixPermissionsCommand.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        $this->databaseDefault   = $databaseDefault;
+
+        parent::__construct();
+    }
+
+    /**
+     * Execute the console command.
+     *
+     * @return int|null
```
Comment (GrahamCampbell) [inline src/Console/Commands/FixPermissionsCommand.php]: should be just `void`

Comment (jbrooksuk) [inline src/Console/Commands/FixPermissionsCommand.php]: Good catch.

File: src/Providers/ConsoleServiceProvider.php [hunk 7]
```diff
@@ -2,7 +2,9 @@
 
 namespace CachetHQ\Cachet\Providers;
 
+use CachetHQ\Cachet\Console\Commands\FixPermissionsCommand;
 use CachetHQ\Cachet\Console\Commands\OneClickDeployCommand;
+use Illuminate\Support\Facades\Config;
```
Comment (GrahamCampbell) [inline src/Providers/ConsoleServiceProvider.php]: Don't use the config facade in here please.

File: src/Providers/ConsoleServiceProvider.php [hunk 8]
```diff
@@ -24,6 +27,20 @@ public function boot()
      */
     public function register()
     {
+        $this->app->singleton('CachetHQ\Cachet\Console\Commands\FixPermissionsCommand', function ($app) {
+            $storageDirectory = $app->path['storage'];
+            $databaseDirectory = $app->path['database'];
+            $databasePath = Config::get('database.connections.sqlite.database');
```
Comment (GrahamCampbell) [inline src/Providers/ConsoleServiceProvider.php]: `$app->config->get(...);`

File: src/Providers/ConsoleServiceProvider.php [hunk 9]
```diff
@@ -24,6 +27,20 @@ public function boot()
...
+            $databaseDefault = Config::get('database.default');
+
+            return new FixPermissionsCommand(
```
Comment (GrahamCampbell) [inline src/Providers/ConsoleServiceProvider.php]: does this fit on one line?

Comment (jbrooksuk) [inline src/Providers/ConsoleServiceProvider.php]: < 120?

Comment (GrahamCampbell) [inline src/Providers/ConsoleServiceProvider.php]: That's a soft limit. If we don't use one line, there's no point in doing the variable asignments above.

File: src/Console/Commands/FixPermissionsCommand.php (hunk 5 above)
Comment (jbrooksuk) [inline src/Console/Commands/FixPermissionsCommand.php]: Lol, old habits.

**Comment by jbrooksuk** [comment]
I can rebase and merge.

File: src/Providers/ConsoleServiceProvider.php [hunk 10]
```diff
@@ -24,6 +26,20 @@ public function boot()
      */
     public function register()
     {
+        $this->app->singleton('CachetHQ\Cachet\Console\Commands\FixPermissionsCommand', function ($app) {
+            $storageDirectory = $app->path['storage'];
+            $databaseDirectory = $app->path['database'];
+            $databasePath = $app->config->get('database.connections.sqlite.database');
+            $databaseDefault = $app->config->get('database.default');
+
+            return new FixPermissionsCommand(
```
Comment (GrahamCampbell) [inline src/Providers/ConsoleServiceProvider.php]: one line plz?

Comment (jbrooksuk) [inline src/Providers/ConsoleServiceProvider.php]: I'm not rebasing again ;) Pushed on top.

**Comment by GrahamCampbell** [comment]
:+1:

**Comment by jbrooksuk** [comment]
Learnt a lot with this, thanks @GrahamCampbell

**Comment by joecohens** [comment]
Awesome!

**Comment by joecohens** [comment]
Thanks to both :smile: :smile: :smile:  !!

**Pull request merged**